app/src/main/java/com/github/drawyourpath/bootcamp/userProfileCreation/UserNameTestAndSetFragment.kt

```diff
@@ -0,0 +1,111 @@
+
+        testUserNameButton.setOnClickListener {
+            usernameAvaibility(database, inputUserName.text.toString(), showTestResult)
+        }
+
+        val setUserNameButton: Button =
+            view.findViewById(R.id.setUserName_button_userProfileCreation)
+        setUserNameButton.setOnClickListener {
+            //try to set the userName to the database
+            val usernameStr = inputUserName.text.toString()
+            val testUsername = usernameAvaibility(database, usernameStr, showTestResult)
+            val previousActivity = activity
+            if (previousActivity != null && testUsername) {
+                database.setUserName(usernameStr)
+                val fragManagement = previousActivity.supportFragmentManager.beginTransaction()
+                val dataToPersoInfoFrag: Bundle = Bundle()
+                //data to transmit to the PersonalInfoFragment(username + isTest)
+                dataToPersoInfoFrag.putBoolean("isRunningTestForDataBase", isTest)
+                dataToPersoInfoFrag.putString("userName", inputUserName.text.toString())
+                val persoInfoFrag = PersonalInfoFragment()
+                persoInfoFrag.arguments = dataToPersoInfoFrag
+                fragManagement.replace(R.id.userName_frame, persoInfoFrag).commit()
+            }
+        }
+    }
+}
+
+/**
+ * Helper function that display a message with outputMessage to the user on the UI and return a boolean that indicate
+ * if the username is available
+ * @param username username with the availability to tes
+ * @param database used to test if the username is available
+ * @param outputMessage editText used to show an error message to the user on the UI if the username is not available
+ * @return true if the username is available, and false otherwise
+ */
+private fun usernameAvaibility(database: Database, username: String, outputMessage: TextView): Boolean{
+    if (username == "") {
+        outputMessage.text = buildString { append("The username can't be empty !") }
+        outputMessage.setTextColor(Color.RED)
+        return false
+    }
+
+    val unAvailableOutput: String = buildString {
+        append("*The username ")
+        append(username)
+        append(" is NOT available !")
+    }
+    val availableOutput: String = buildString {
+        append("*The username ")
+        append(username)
+        append(" is available !")
+    }
+    val future = database.isUserNameAvailable(username)
+
+    //since the orTimeout require an API level 33(we are in min API level 28), we can't use it
+    val durationFuture = future.thenAccept {
+        if (it) {
+            outputMessage.text = availableOutput
+            outputMessage.setTextColor(Color.GREEN)
+        } else {
+            outputMessage.text = unAvailableOutput
+            outputMessage.setTextColor(Color.RED)
+        }
+    }
+    if(outputMessage.text.toString() == availableOutput){
```

Comment:
We still have a race condition here. If we really want the boolean result we should use a future.

Reply:
I'm sorry but I don't arrive to do this method like your advice. I can't have access to a lot of timeout future function since they are only available on API 33. That's the best way I can do. If you have an idea, we can discuss it.

Reply:
The returned boolean by the method is independent from the actual availability of the username, so we might as well just remove the return value for now.

The thing is that you can't return synchronously a value that depends on asynchronous operations so even with a timeout function, it wouldn't be possible to return a boolean matching the result. It must be a future.

Reply:
I have check with multiple case and this block the possibility to the user to set a username that is already taken. So if I removed the return nothing, the username can be added to the database. And we must not do that .

Reply:
I correct it based on your remark with the help of @HoLuDyEr. Tank you